subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenSnapshotResolveIntegrationTest.groovy

```diff
@@ -1075,6 +1075,90 @@ class CheckIsChangingRule implements ComponentMetadataRule {
         false       | Sources.POM
     }
 
+    @RequiredFeature(feature = GradleMetadataResolveRunner.REPOSITORY_TYPE, value = "maven")
+    def "can resolve files from new variant with unique and non unique snapshots"() {
+        given:
+        buildFile << """
+dependencies.components {
+    all(PomRule) {
+        params("runtime")
+    }
+    all(PomRule) {
+        params("runtimeElements")
+    }
+}
+
+dependencies {
+    compile("org.gradle.integtests.resolve:unique:1.0-SNAPSHOT")
+    compile("org.gradle.integtests.resolve:not-unique:1.0-SNAPSHOT")
+}
+
+configurations {
+    poms {
+        extendsFrom(configurations.compile)
+        canBeResolved = true
+        canBeConsumed = false
+        attributes {
+            attribute(Category.CATEGORY_ATTRIBUTE, getObjects().named(Category, Category.DOCUMENTATION))
+            attribute(DocsType.DOCS_TYPE_ATTRIBUTE, getObjects().named(DocsType, "pom"))
+        }
+    }
+}
+
+tasks.register("downloadPoms", Copy) {
+    from(configurations.poms)
+    into("\$buildDir/poms")
+}
+
+abstract class PomRule implements ComponentMetadataRule {
+
+    final String baseVariant
+
+    @Inject
+    PomRule(String baseVariant) {
+        this.baseVariant = baseVariant
+    }
+
+    @Inject
+    abstract ObjectFactory getObjects()
+
+    void execute(ComponentMetadataContext context) {
+        context.details.maybeAddVariant("pom", baseVariant) {
+            attributes {
+                attribute(Category.CATEGORY_ATTRIBUTE, getObjects().named(Category, Category.DOCUMENTATION))
+                attribute(DocsType.DOCS_TYPE_ATTRIBUTE, getObjects().named(DocsType, "pom"))
+            }
+            withFiles {
+                removeAllFiles()
+                addFile("\${context.details.id.name}-\${context.details.id.version}.pom")
```

Comment:
Minor suggestion, feel free to ignore: if it was a black box to test, it could also be worth it to test how this works for files that Gradle is usually not aware of, unlike a POM which should be definitely there and even already downloaded at this point.

Reply:
The POM file downloaded that way is handled as an artifact, not as the metadata file. Gradle currently does not optimize this and will download and cache the file in different locations as metadata files are handled specially.

So the test is valid.